### PR TITLE
Improve behaviour for playing from history

### PR DESCRIPTION
**Pull request by chme**
This is an attempt at improving the selection of a song from the "previously played" list. 

Remote sends the command "playnow" with mode "-1" if an entry of the history was selected. The index parameter is then the position of this entry in the history queue. Currently the mode parameter is not taken into account and therefor issuing this action results in playing the song from the "up next" queue at the position identified by the index parameter.

With these changes if mode is "-1", forked-daapd will look in the history to identify the song to start playing.
If the song is found and it is also present in the playqueue, forked-daapd will skip back to this song in the playqueue (gets the song id from the history and starts playing this id).

What is not working after these changes is, that if a song is not in the current playqueue, the queue will start from the beginning. Another problem is, if a song is in the playqueue more than once. forked-daapd will skip to the first occurrence in the playqueue.

### CODE REVIEW

File: src/httpd_dacp.c [hunk 1]
```diff
@@ -783,12 +785,42 @@ dacp_reply_cue_play(struct evhttp_request *req, struct evbuffer *evbuf, char **u
...
-  /* If selection was from Up Next queue (command will be playnow), then index is relative */
-  //TODO playnow for mode = 1 (index is relativ to the history queue
+  /* If selection was from Up Next queue or history queue (command will be playnow), then index is relative */
+  hist = 0;
   if ((param = evhttp_find_header(query, "command")) && (strcmp(param, "playnow") == 0))
-    pos += status.pos_pl;
+    {
+      /* If mode parameter is -1, the index is relative to the history queue, otherwise to the Up Next queue */
+      param = evhttp_find_header(query, "mode");
+      if (param && strcmp(param, "-1") == 0)
```
Comment (ejurgensen) [inline src/httpd_dacp.c]: Could you add parenthesis, so: (strcmp(param, "-1") == 0)

File: src/httpd_dacp.c [hunk 2]
```diff
@@ -783,12 +785,42 @@ dacp_reply_cue_play(struct evhttp_request *req, struct evbuffer *evbuf, char **u
...
+	{
+	  /* Play from history queue */
+	  hist = 1;
+	  history = player_history_get();
+	  if (history->count > pos)
+	    {
+	      pos = history->id[(history->start_index + history->count - pos - 1) % MAX_HISTORY_COUNT];
```
Comment (ejurgensen) [inline src/httpd_dacp.c]: Could you split this in two lines, so first calculate the id and then find the pos?

**Comment by chme** [comment]
Updated and rebased. I have to admit that the code is really difficult to read and follow. But refactoring it will lead to a lot of changes with the potential of breaking things. But if you have some ideas how to improve things, i can try to do it.

**Comment by ejurgensen** [comment]
Thanks for this. I don't really have any ideas for simplification, but yes, I have trouble following the code too.

**Pull request merged**